From 71f5ba1c9e041490ebf9b0687ab9798cd7517691 Mon Sep 17 00:00:00 2001
From: dann frazier <dann.frazier@chainguard.dev>
Date: Tue, 15 Sep 2026 11:33:31 -0600
Subject: [PATCH 03/18] Revert "printenv: don't quote output when given
 arguments and redirected"

This reverts commit ea1c2b178777b6e0033daa4c1e77f667eaae5f70.
NEWS left untouched.
---
 doc/coreutils.texi        | 21 ++++++++-------------
 src/printenv.c            |  5 ++---
 tests/misc/printenv.sh    | 18 +++++-------------
 tests/misc/tty-quoting.sh | 14 ++++----------
 4 files changed, 19 insertions(+), 39 deletions(-)

diff --git a/doc/coreutils.texi b/doc/coreutils.texi
index 46854db4a..9fe4ba032 100644
--- a/doc/coreutils.texi
+++ b/doc/coreutils.texi
@@ -15519,15 +15519,14 @@ The program accepts the following option.  Also see @ref{Common options}.
 
 @end table
 
-If no @var{variable}s are specified, @command{printenv} quotes
-environment variables and their values using the @samp{shell-escape}
-style.  If @var{variable}s are specified, @command{printenv} only quotes
-their values using the @samp{shell-escape} style if standard output is a
-terminal.
-
-The environment variable @env{QUOTING_STYLE}
-specifies the quoting style.  Valid quoting styles are:
+@macro printenvAlwaysQuoted
+The printed environment variables and their values are quoted using the
+@samp{shell-escape} style.   The environment variable
+@env{QUOTING_STYLE} specifies the quoting style.  Valid quoting styles
+are:
 @quotingStyles
+@end macro
+@printenvAlwaysQuoted
 
 @cindex exit status of @command{printenv}
 Exit status:
@@ -17497,11 +17496,7 @@ If no command name is specified following the environment
 specifications, the resulting environment is printed.  This is like
 specifying the @command{printenv} program.
 
-The printed environment variables and their values are quoted using the
-@samp{shell-escape} style.   The environment variable
-@env{QUOTING_STYLE} specifies the quoting style.  Valid quoting styles
-are:
-@quotingStyles
+@printenvAlwaysQuoted
 
 For some examples, suppose the environment passed to @command{env}
 contains @samp{LOGNAME=rms}, @samp{EDITOR=emacs}, and
diff --git a/src/printenv.c b/src/printenv.c
index dba87f373..540f56334 100644
--- a/src/printenv.c
+++ b/src/printenv.c
@@ -111,9 +111,8 @@ main (int argc, char **argv)
     }
 
   bool quote_output = false;
-  idx_t const n_args = argc - optind;
 
-  if (!opt_nul_terminate_output && (n_args <= 0 || isatty (STDOUT_FILENO)))
+  if (!opt_nul_terminate_output)
     {
       int qs = getenv_quoting_style ();
       if (qs < 0)
@@ -131,7 +130,7 @@ main (int argc, char **argv)
   if (opt_nul_terminate_output)
     xset_binary_mode (STDOUT_FILENO, O_BINARY);
 
-  if (n_args <= 0)
+  if (optind >= argc)
     {
       for (char **env = environ; *env != NULL; ++env)
         print_envvar (*env, terminator, quote_output);
diff --git a/tests/misc/printenv.sh b/tests/misc/printenv.sh
index a6eab09f6..8ec228a6a 100755
--- a/tests/misc/printenv.sh
+++ b/tests/misc/printenv.sh
@@ -84,21 +84,17 @@ compare /dev/null out || fail=1
 cat <<\EOF >exp-noargs-literal || framework_failure_
 a b=c d
 EOF
-cat <<\EOF >exp-arg-literal || framework_failure_
-c d
-EOF
 cat <<\EOF >exp-args-literal || framework_failure_
 c d
-c d
 EOF
 cat <<\EOF >exp-noargs-shell || framework_failure_
 'a b'='c d'
 EOF
-cp exp-arg-literal exp-arg-shell &&
-cp exp-args-literal exp-args-shell || framework_failure_
-for t in noargs arg args; do
-  tr "'" '"' <exp-$t-shell >exp-$t-c || framework_failure_
-done
+cat <<\EOF >exp-args-shell || framework_failure_
+'c d'
+EOF
+tr "'" '"' <exp-noargs-shell >exp-noargs-c || framework_failure_
+tr "'" '"' <exp-args-shell >exp-args-c || framework_failure_
 for qs in literal shell c; do
   env -i PATH="$PATH" QUOTING_STYLE=$qs 'a b'='c d' \
     printenv >out-t 2>err || fail=1
@@ -108,10 +104,6 @@ for qs in literal shell c; do
   compare /dev/null err || fail=1
   env -i PATH="$PATH" QUOTING_STYLE=$qs 'a b'='c d' \
     printenv 'a b' >out 2>err || fail=1
-  compare exp-arg-$qs out || fail=1
-  compare /dev/null err || fail=1
-  env -i PATH="$PATH" QUOTING_STYLE=$qs 'a b'='c d' \
-    printenv 'a b' 'a b' >out 2>err || fail=1
   compare exp-args-$qs out || fail=1
   compare /dev/null err || fail=1
 done
diff --git a/tests/misc/tty-quoting.sh b/tests/misc/tty-quoting.sh
index 833dfaedb..a0dc1d4b8 100755
--- a/tests/misc/tty-quoting.sh
+++ b/tests/misc/tty-quoting.sh
@@ -44,22 +44,16 @@ run_tty_ env test -t 1 ||
 run_tty_ env printf foo >printf.t &&
  skip_ 'libc buffering induced a tty probe'
 
-for cmd in basename du dirname 'ls -w0' readlink 'realpath --relative-to=.' \
-           printenv; do
+for cmd in basename du dirname 'ls -w0' readlink 'realpath --relative-to=.'; do
   test "$cmd" = 'du' && field=2 || field=1
-  case "$cmd" in
-    dirname) file='f oo/.' ;;
-    printenv) file=TEST_ENV1 ;;
-    *) file='f oo' ;;
-  esac
+  test "$cmd" = 'dirname' && file='f oo/.' || file='f oo'
 
-  TEST_ENV1='f oo' run_tty_ $cmd "$file" >quoted.t || fail=1
+  run_tty_ $cmd "$file" >quoted.t || fail=1
   cut -f$field- quoted.t >quoted || framework_failure_
 
   # Note ls theoretically doesn't need isatty() for a specified QUOTING_STYLE
   # but it does need it to determine appropriate output format.
-  TEST_ENV1='f oo' QUOTING_STYLE=literal run_tty_ $cmd "$file" \
-    >unquoted.t || fail=1
+  QUOTING_STYLE=literal run_tty_ $cmd "$file" >unquoted.t || fail=1
   cut -f$field- unquoted.t >unquoted || framework_failure_
 
   env printf '%q\n' "$(cat unquoted)" >printf_quoted || framework_failure_
-- 
2.53.0

